translated/tech/20190707-go-unknown-parts-of-the-test-package.md

```diff
@@ -0,0 +1,270 @@
+PASS
+ok   math 0.007s
+```
+
+再执行一次的话缓存就会生效了。缓存是测试内容、环境变量和命令行参数的哈希。一旦计算出来，这个缓存会转储到`$GOCACHE`指向的文件夹（Unix系统下默认实`$XDG_CACHE_HOME`或`$HOME/.cache`）。清空这个文件夹也就会清空缓存。
+
+关于标识符的话，如[文档](https://golang.org/cmd/go/#hdr-Test_packages)所述，并不是所有标识符都是可缓存的：
+
+> 缓存一致的定义为：测试涉及的二进制可执行文件一样，同时命令行标识符属于'可缓存的'测试标识符限定子集（包括`-cpu`，`-list`，`parallel`，`-run`，`-short`和`-v`等）。使用任何不属于可缓存范围的标识符或参数都会导致缓存失效。显式屏蔽缓存的习惯用法是采用`-count=1`标识符。
+
+因为`count`规定测试必须执行的次数，因此`-count=1`显式地声明测试应该不多不少地只运行一次，使得这个标识符成为规避缓存的最优习惯用法。
+
+再提一下：Go1.12之前可通过设置[`GOCACHE`](https://golang.org/doc/go1.12#gocache)环境变量`GOCACHE=off go test math/`的方式绕过缓存。
+
+运行测试时，Go会逐个包依次运行它们。Go处理测试包名的方式也给测试提供了更多策略。
+
+## 白盒测试 vs 黑盒测试
+
+黑盒测试不触及内部结构，只能访问到导出的函数和结构，而白盒测试则允许深入到非导出函数的内部实现。两种方式都是Go原生支持的。以下是展示黑盒测试和白盒测试各自优势的简单程序：
+
+```go
+package deck
+
+import (
+	"errors"
+	"math/rand"
+)
+
+var Empty = errors.New("Empty deck")
+
+type Deck struct {
+	cards    []uint8
+	shuffled bool
+}
+
+func NewDeck(numbers uint8) *Deck {
+	cards := make([]uint8, 0, numbers)
+	for i := uint8(0); i < numbers; i++ {
+		cards = append(cards, i+1)
+	}
+
+	d := Deck{cards: cards}
+
+	return &d
+}
+
+func (d *Deck) Draw() (card uint8, err error) {
+	if !d.shuffled {
+		d.shuffle()
+	}
+
+	if len(d.cards) == 0 {
+		return 0, Empty
+	}
+	card, d.cards = d.cards[0], d.cards[1:]
+
+	return card, nil
+}
+
+func (d *Deck) shuffle() {
+	rand.Shuffle(len(d.cards), func(i, j int) {
+		d.cards[i], d.cards[j] = d.cards[j], d.cards[i]
+	})
+	d.shuffled = true
+}
+```
+
+> 原文给出的`deck`包的`import`路径不对，已修正如上
+
+上述代码只是执行洗牌操作后让用户抽牌。黑盒测试确保牌组能够创建并依次抽取直到没牌。
+
+```go
+package deck_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"deck"
+)
+
+func TestDeckCanDrawCards(t *testing.T) {
+	var num uint8 = 10
+
+	d := deck.NewDeck(num)
+	for i := uint8(0); i < num; i++ {
+		_, err := d.Draw()
+		assert.Nil(t, err)
+	}
+	_, err := d.Draw()
+	assert.Equal(t, err, deck.Empty)
+}
+```
+
+编写黑盒测试的唯一要求是给包名加上`_test`后缀。这个包被看作不同于`deck`的包，所以无法访问到非导出的函数。Go原生支持这个方式，编译器不会抱怨同一个文件夹下有两个不同包名。
+
+白盒测试则检验牌组只会在第一次抽取时被洗一次：
+
+```go
+package deck
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDeckShouldBeShuffledOnce(t *testing.T) {
+	var num uint8 = 5
+
+	d := NewDeck(num)
+	assert.Equal(t, len(d.cards), int(num))
+	assert.Equal(t, d.shuffled, false, "Deck should init as not shuffled")
+	orderBefore := fmt.Sprint(d.cards)
+
+	d.shuffle()
+	assert.Equal(t, d.shuffled, true, "Deck has not been marked as shuffled")
+	orderAfter := fmt.Sprint(d.cards)
+
+	assert.NotEqual(t, orderBefore, orderAfter, "Deck once shuffled should have new card order")
+}
+```
+
+上述测试采用和`deck`相同的包名，因此能够访问到非导出的函数。
+
+但是，白盒测试有这么一个短板。黑盒测试导出的函数能够保证正确的结果不受包的内部实现影响。因此，我们能够自由地改变和优化内部实现而不破坏现有测试。而对于白盒测试，它是和内部实现绑定的，优化操作有可能会破坏现有测试。
+
+接下来看一下这个测试包的其他特性：性能测试。
+
+## 只执行一次性能测试
+
+[Go 1.12](https://golang.org/doc/go1.12#testing)引入的`-benchtime=1x`、`-benchtime=10x`等允许性能测试只执行我们想要的次数。这个`-benchtime=1x`标识符为测试提供了便利，因为它使得我们只需运行至少一次性能测试就可验证上一次变更是否和现有代码兼容。
```

Comment:
The flag -benchtime=1x can be useful with your tests suite since it would allow you to run your benchmarks at least once in order to check that they are not broken due to your last change.
标志 -benchtime = 1x 对测试套件（test suite）十分有用，因为他允许你运行性能测试最低一次，以检测上次代码变更是否破坏现有测试

这句话感觉翻译变动较大~~ test suite 是个软件术语

Reply:
- 那 test suite 不译
- 整句话更新为“这个`-benchtime=1x`标识符对测试套件（test suite）是很有用的，它使得我们只需运行至少一次性能测试即可验证上一次变更是否破坏了现有代码。”

Reply:
这个可以，比较通顺